### PR TITLE
Add stop_ongoing_execution flag to rebalance requests for full run

### LINKED ISSUE #10571: Refreshing a remove brokers operation on `KafkaRebalance` resource while one rebalancing is already running can drive to a `NotReady` state
Create a `Kafka` custom resource (for example with 7 brokers) with the `cruiseControl` field to run Cruise Control within the cluster deployment.
Run a rebalancing by creating a `KafkaRebalance` custom resource to remove nodes 5, 6 (with auto-approval enabled), like this:

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaRebalance
metadata:
  name: my-rebalance
  labels:
    strimzi.io/cluster: my-cluster
  annotations:
    strimzi.io/rebalance-auto-approval: "true"
# no goals specified, using the default goals from the Cruise Control configuration
spec:
  mode: remove-brokers
  brokers: [5, 6]
```

Wait for the rebalancing to go from ProposalPendy, to ProposalReady and automatically (auto-approval enabled) to Rebalancing.
While rebalancing is running, ask for a new rebalancing (using the "refresh" annotation on the already existing custom resource) including nodes 3, 4 as well, so having all 3,4,5 and 6, like this:

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaRebalance
metadata:
  name: my-rebalance
  labels:
    strimzi.io/cluster: my-cluster
  annotations:
    strimzi.io/rebalance-auto-approval: "true"
    strimzi.io/rebalance: "refresh"
# no goals specified, using the default goals from the Cruise Control configuration
spec:
  mode: remove-brokers
  brokers: [3, 4, 5, 6]
```

Sometimes (it could depending on the timing and where Cruise Control is on the current rebalancing), the operator will go through the following log error and the `KafkaRebalance` moves to `NotReady` state:

```shell
2024-09-12 13:51:10 INFO  KafkaRebalanceAssemblyOperator:1113 - Reconciliation #59(watch) KafkaRebalance(myproject/my-rebalance): Requesting Cruise Control rebalance [dryrun=true]
2024-09-12 13:51:11 INFO  KafkaRebalanceAssemblyOperator:351 - Reconciliation #59(watch) KafkaRebalance(myproject/my-rebalance): KafkaRebalance state is now updated to [ProposalReady] with annotation strimzi.io/rebalance=refresh applied on the KafkaRebalance resource
2024-09-12 13:51:11 INFO  KafkaRebalanceAssemblyOperator:359 - Reconciliation #59(watch) KafkaRebalance(myproject/my-rebalance): Removing annotation strimzi.io/rebalance=refresh
2024-09-12 13:51:11 INFO  AbstractOperator:520 - Reconciliation #60(watch) KafkaRebalance(myproject/my-rebalance): KafkaRebalance my-rebalance in namespace myproject was MODIFIED
2024-09-12 13:51:11 INFO  AbstractOperator:520 - Reconciliation #61(watch) KafkaRebalance(myproject/my-rebalance): KafkaRebalance my-rebalance in namespace myproject was MODIFIED
2024-09-12 13:51:11 INFO  CrdOperator:123 - Reconciliation #59(watch) KafkaRebalance(myproject/my-rebalance): Status of KafkaRebalance my-rebalance in namespace myproject has been updated
2024-09-12 13:51:11 INFO  AbstractOperator:546 - Reconciliation #59(watch) KafkaRebalance(myproject/my-rebalance): reconciled
2024-09-12 13:51:11 INFO  AbstractOperator:266 - Reconciliation #60(watch) KafkaRebalance(myproject/my-rebalance): KafkaRebalance my-rebalance will be checked for creation or modification
2024-09-12 13:51:11 INFO  KafkaRebalanceAssemblyOperator:317 - Reconciliation #60(watch) KafkaRebalance(myproject/my-rebalance): Rebalance action is performed and KafkaRebalance resource is currently in [ProposalReady] state
2024-09-12 13:51:11 INFO  KafkaRebalanceAssemblyOperator:788 - Reconciliation #60(watch) KafkaRebalance(myproject/my-rebalance): Auto-approval set on the KafkaRebalance resource
2024-09-12 13:51:11 INFO  KafkaRebalanceAssemblyOperator:1113 - Reconciliation #60(watch) KafkaRebalance(myproject/my-rebalance): Requesting Cruise Control rebalance [dryrun=false]
2024-09-12 13:51:11 ERROR KafkaRebalanceAssemblyOperator:378 - Reconciliation #60(watch) KafkaRebalance(myproject/my-rebalance): Status updated to [NotReady] due to error: Error for request: my-cluster-cruise-control.myproject.svc:9090/kafkacruisecontrol/remove_broker?json=true&dryrun=false&verbose=true&skip_hard_goal_check=false&brokerid=3%2C4%2C5%2C6. Server returned: Error processing POST request '/remove_broker' due to: 'java.lang.IllegalStateException: Cannot start a new execution while there is an ongoing execution. Please use stop_ongoing_execution=true to stop ongoing execution and start a new one.'.
```

It seems that asking for a new rebalancing with different nodes to remove needs that currently running task is stopped via `stop_ongoing_execution=true` in the query string on the POST request to the REST API.
Maybe we should have this addition in any POST operation for rebalancing when our intention is to not waiting for the current operation ending but starting a new one straight away.

### PR DESCRIPTION
**Pull request by tinaselenge**
### Type of change

_Select the type of your PR_

- Bugfix

### Description

This adds an option to set stop_ongoing_execution flag. 
The new flag is set when sending rebalancing requests for full run to avoid "Cannot start a new execution while there is an ongoing execution" error. This error can happen even if the operator calls CC's endpoint `/stop_proposal_execution` before sending a new rebalance request. The reason for this is documented [here](https://github.com/linkedin/cruise-control/wiki/rest-apis) as:
```
Note that Cruise Control does not wait for the ongoing batch to finish when it stops execution, i.e. the in-progress batch may still be running after Cruise Control stops the execution.
```

Here is the possible flow that causes this issue:
1. Currently KafkaRebalance CR for removing brokers is in Rebalancing state
2. User updates the CR with a refresh annotation and with a new set of brokers to remove
3. The operator sends a request to `/stop_proposal_execution` endpoint to stop the current rebalance operation
4. The request completes successfully, however there are still in-progress batch for the current balance operation in CC.
5. The operator sends a request for a new proposal for the updated set of brokers to remove. 
6. The new proposal is ready, therefore the KafkaRebalance is in ProposalReady state
7. The operator sends a request to execute the removal of the updated set of brokers
8. This request fails because there are still in-progress batch of the previous rebalance operation.
9. KafkaRebalance is in NotReady state due to this failure. 

The fix is to update step 7, to send the request with "stop_ongoing_execution" flag set to true, so that CC first stops the still in-progress batch of the previous rebalance operation before processing the new rebalance request. The flow would become:

  7. The operator sends a request to execute the removal of the updated set of brokers with "stop_ongoing_execution" flag set to true
  8. CC waits for the still in-progress batch of the previous rebalance operation to stop
  9. CC processes the request to execute the removal of the updated set of brokers
  10. The KafkaRebalance is in Ready.

There were no test cases for the existing flow where operator calls `/stop_proposal_execution` when KafkaRebalance is in Rebalancing state and then refresh annotation is applied. These missing tests cases are added , however they do not test the new flow because we cannot mock the internal state of CC to have still in-progress batch. 

Closes #10571

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

### CODE REVIEW

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java [hunk 1]
```diff
@@ -862,7 +862,7 @@ private Future<MapAndStatus<ConfigMap, KafkaRebalanceStatus>> onProposalReady(Re
                     return configMapOperator.getAsync(kafkaRebalance.getMetadata().getNamespace(), kafkaRebalance.getMetadata().getName()).compose(loadmap -> Future.succeededFuture(new MapAndStatus<>(loadmap, buildRebalanceStatusFromPreviousStatus(kafkaRebalance.getStatus(), StatusUtils.validate(reconciliation, kafkaRebalance)))));
                 case approve:
                     LOGGER.debugCr(reconciliation, "Annotation {}={}", ANNO_STRIMZI_IO_REBALANCE, KafkaRebalanceAnnotation.approve);
-                    return requestRebalance(reconciliation, host, apiClient, kafkaRebalance, false, rebalanceOptionsBuilder);
+                    return requestRebalance(reconciliation, host, apiClient, kafkaRebalance, false, rebalanceOptionsBuilder, true);
```
Comment (ppatierno) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:865]: you are using the stop ongoing execution only when we are approving a proposal ... what about refreshing while it's in ProposalPending or Rebalancing? (of course refreshing in ProposalReady doesn't make sense because there is nothing going on).

**Review by ppatierno — COMMENTED** [review]
Mostly good, I left one comment.
Other than that, could you please add one more test in the `KafkaRebalanceStateMachineTest` class please?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java (hunk 1 above)
Comment (scholzj) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:865]: Why do we need to stop ongoing execution when approving a proposal in the first place?

Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:865]: Good point ... what's the use case you saw @tinaselenge ?

**Comment by ppatierno** [comment]
@tinaselenge could you rebase against main, I merged another PR fixing some reconciliation issues on KafkaRebalance.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java (hunk 1 above)
Comment (tinaselenge) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:865]: So this is my understanding of the code based on my recreation of the problem, please correct me if I'm wrong. The issue mainly happens when you refresh, while the KafkaRabalance is in Rebalancing state (let's say rebalance_1 is in progress). When refresh annotation applied, we send a request to stop the ongoing rebalance operation, however this does not stop rebalance_1 completely, and that's the problem. Immediately after the stop request completes, we send a request for a new proposal (dry_run=true), then the state becomes ProposalReady. If auto-approval is set or user manually approved this new proposal, we send a new rebalance request (dry_run=false), let's call it rebalance_2. However the request for rebalance_2 fails if the rebalance_1 is still in progress. This change makes sure that rebalance_1 is completely stopped and then the request for rebalance_2 is processed. I tested this only with auto-approval annotation, however I think it is possible to happen with manual approval by user, if they approved it quickly and rebalance_1 was taking a long time. 

I don't think we need this flag set for ProposalPending or Rebalancing states, because when refresh annotation is applied while on these 2 states, we only send a request for new proposal (dry_run=false). We only need this flag set when we send dry_run=true requests.

**Comment by tinaselenge** [comment]
Thanks for the review!
@ppatierno  I have added a test case to KafkaRebalanceStateMachineTest and then rebased against main.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java (hunk 1 above)
Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:879]: thanks @tinaselenge what you described is what I faced during auto-rebalance where "auto-approval" is set and people can ask two consecutive scale up/down which result in two consecutive rebalancing requests.

> I don't think we need this flag set for ProposalPending or Rebalancing states, because when refresh annotation is applied while on these 2 states, we only send a request for new proposal (dry_run=false). We only need this flag set when we send dry_run=true requests.

Did you mean the other way around?  I mean "we only send a request for new proposal (dry_run=**true**). We only need this flag set when we send dry_run=**false** requests."

Because in `PendingProposal` we obviously ask for a new ... proposal with dry_run=true not false.
Anyway even when we ask for a new proposal, while a proposal is still processing, we should be sure that CC doesn't reply the same way so we should stop execution anyway (maybe not, because it's different from an actual rebalance, but it's worth checking on CC codebase). The same applied in `Rebalancing` where we ask for a new proposal (dry_run=true) is refresh is applied.

Long story short, the question is, is stop going execution flag valid only when you run an actual rebalance or even when you asked to CC to process a proposal?

Comment (tinaselenge) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:879]: > Did you mean the other way around? I mean "we only send a request for new proposal (dry_run=true). We only need this flag set when we send dry_run=false requests."

Yes, I meant the other way around. Sorry!

Comment (tinaselenge) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:879]: >Because in PendingProposal we obviously ask for a new ... proposal with dry_run=true not false.
Anyway even when we ask for a new proposal, while a proposal is still processing, we should be sure that CC doesn't reply the same way so we should stop execution anyway (maybe not, because it's different from an actual rebalance, but it's worth checking on CC codebase). The same applied in Rebalancing where we ask for a new proposal (dry_run=true) is refresh is applied.

I agree that we should stop the ongoing execution when we ask for a new proposal due to "refresh" annotation so the optimisation calculation is up to date. And we do already have this logic, if you see the "Rebalancing()" function.  In this function, the operator first calls the stop endpoint and then requests a new proposal. The issue is sometimes CC internally can still have in progress batch, even though stop endpoint was called. I can spend sometime looking into CC code.

Comment (tinaselenge) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:879]: > Long story short, the question is, is stop going execution flag valid only when you run an actual rebalance or even when you asked to CC to process a proposal?

It seems to be effective when we run an actual rebalance, not when we asked for a proposal.

**Comment by tinaselenge** [comment]
I also updated the PR description with more details.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java (hunk 1 above)
Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:879]: > The issue is sometimes CC internally can still have in progress batch, even though stop endpoint was called.

So it means that even if we call the `stopExecution` we should then call the `requestRebalance` with the stop_ongoing_execution=true even in the `onRebalancing`. At this point I was wondering if we should actually totally remove the `stopExecution` call and just pass the corresponding flag.

Comment (tinaselenge) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:879]: Could it be still beneficial to have the stopExecution call because it gets called before requesting a new proposal?The flag is set in a later call to request an actual rebalance operation but the existing rebalance might be already stopped by the earlier stopExecution, and in the case it's not, the flag helps.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java [hunk 2]
```diff
@@ -1129,22 +1129,33 @@ private boolean isKafkaClusterReady(Kafka kafka) {
...
+
     private Future<MapAndStatus<ConfigMap, KafkaRebalanceStatus>> requestRebalance(Reconciliation reconciliation,
                                                           String host, CruiseControlApi apiClient, KafkaRebalance kafkaRebalance,
                                                           boolean dryrun, AbstractRebalanceOptions.AbstractRebalanceOptionsBuilder<?, ?> rebalanceOptionsBuilder) {
-        return requestRebalance(reconciliation, host, apiClient, kafkaRebalance, dryrun, rebalanceOptionsBuilder, null);
+        return requestRebalance(reconciliation, host, apiClient, kafkaRebalance, dryrun, rebalanceOptionsBuilder, null, false);
     }
 
 
     private Future<MapAndStatus<ConfigMap, KafkaRebalanceStatus>> requestRebalance(Reconciliation reconciliation, String host, CruiseControlApi apiClient, KafkaRebalance kafkaRebalance,
```
Comment (ppatierno) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:1146]: I noticed that this method is actually called only from the previous one by providing `userTaskID` as null (so in general it means we are passing `userTaskID` null everywhere). Taking this out, anyway I think that adding one more method is confusing. Can we conflate in just one method with both `userTaskID` parameter and `stopOngoingExecution` and using it? (of course `userTaskID` will be always null as today, we can figure out why later).

Comment (tinaselenge) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:1138]: You are right, adding another method doesn't make sense. I think previously we used to get rebalance status via rebalance endpoint therefore needed to specify user task id. After https://github.com/strimzi/strimzi-kafka-operator/pull/10607, we no longer need it. So I think we should stop passing it to the method.

Comment (tinaselenge) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:1138]: Perhaps, that should be done in a separate PR?

Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:1138]: > Perhaps, that should be done in a separate PR?

yeah I agree and we'll think more through. For now I am happy not having an additional one :-)

**Comment by tinaselenge** [comment]
This might not be the best way to solve this issue, as this change would result in stopping other unrelated executions such as topic reconfiguration whenever we request a full run rebalance (regardless of the refresh annotation). It doesn't seem to be straight forward as adding this flag. Ideally, we would add this flag only when full rebalance request being sent after refresh annotation is applied, however, currently we only send dry run rebalance request after refresh annotation is applied. Once the dry run rebalance is completed, we remove the refresh annotation and start a new reconciliation. In this new reconciliation, we request a full run rebalance but at this point, we don't know if there was a refresh annotation or not, so we can't set the flag based on it.

**Comment by tinaselenge** [comment]
Closing the PR due to https://github.com/strimzi/strimzi-kafka-operator/issues/10571#issuecomment-2443770289. I will open a separate PR for adding tests for the current behaviour of stopping execution via stop endpoint on renew annotation.
